tsfc/coffee_mode.py

```diff
@@ -0,0 +1,282 @@
+    for n in traversal(expressions):
+        if isinstance(n, Failure):
+            return expressions
+
+    def classify(argument_indices, expression):
+        n = len(argument_indices.intersection(expression.free_indices))
+        if n == 0:
+            return OTHER
+        elif n == 1:
+            if isinstance(expression, (Indexed, Conditional)):
+                return ATOMIC
+            else:
+                return COMPOUND
+        else:
+            return COMPOUND
+
+    argument_indices = tuple(itertools.chain(*argument_multiindices))
+    # Apply argument factorisation unconditionally
+    classifier = partial(classify, set(argument_indices))
+    monomial_sums = collect_monomials(expressions, classifier)
+    return [optimise_monomial_sum(ms, argument_indices) for ms in monomial_sums]
+
+
+def index_extent(factor, argument_indices):
+    """Compute the product of the extents of argument indices of a GEM expression
+
+    :arg factor: GEM expression
+    :arg argument_indices: set of argument indices
+
+    :returns: product of extents of argument indices
+    """
+    return numpy.product([i.extent for i in set(factor.free_indices).intersection(argument_indices)])
+
+
+def monomial_sum_to_expression(monomial_sum):
+    """Convert a monomial sum to a GEM expression.
+
+    :arg monomial_sum: an iterable of :class:`Monomial`s
+
+    :returns: GEM expression
+    """
+    indexsums = []  # The result is summation of indexsums
+    # Group monomials according to their sum indices
+    groups = groupby(monomial_sum, key=lambda m: frozenset(m.sum_indices))
+    # Create IndexSum's from each monomial group
+    for _, monomials in groups:
+        # Pick sum indices from the first monomial
+        sum_indices = monomials[0].sum_indices
+        # Create one product for each monomial
+        products = [make_product(monomial.atomics + (monomial.rest,)) for monomial in monomials]
+        indexsums.append(IndexSum(make_sum(products), sum_indices))
+    return make_sum(indexsums)
+
+
+def find_optimal_atomics(monomials, argument_indices):
+    """Find optimal atomic common subexpressions, which produce least number of
+    terms in the resultant IndexSum when factorised.
+
+    :arg monomials: An iterable of :class:`Monomial`s, all of which should have
+                    the same sum indices
+    :arg argument_indices: tuple of argument indices
+
+    :returns: list of atomic GEM expressions
+    """
+    index = itertools.count()  # counter for variables used in ILP
+    atomic_index = OrderedDict()  # Atomic GEM node -> int
+    connections = []
+    # add connections (list of tuples, items in each tuple form a product)
+    for monomial in monomials:
+        connection = []
+        for atomic in monomial.atomics:
+            if atomic not in atomic_index:
+                atomic_index[atomic] = next(index)
+            connection.append(atomic_index[atomic])
+        connections.append(tuple(connection))
```

Comment:
it seems `defaultdict.get(key)` is different from `defaultdict[key]`
[https://alexisylchan.wordpress.com/2012/07/11/python-defaultdict-d-getkey-vs-dkey/](url)